src/UCI.cpp

```diff
@@ -311,8 +311,15 @@ void UCI::loop(const std::string& start) {
           myprintf_so("id name lczero " PROGRAM_VERSION "\nuciok\n");
       }
       else if (token == "setoption")  setoption(is);
-      else if (token == "go")         go(search,bh,is);
-      else if (token == "stop")       search.please_stop();
+      else if (token == "go") {
+          search.please_stop();
+          if (search_thread && search_thread->joinable()) search_thread->join();
+          search_thread = go(search, bh, is);
+      }
+      else if (token == "stop") {
+          search.please_stop();
+          if (search_thread && search_thread->joinable()) search_thread->join();
```

Comment:
You probably want `search_thread = nullptr;` here for good measure (even though it works without it).

Reply:
Could you try it and let me know if it fixes the crashes on your machine?

Reply:
you should also join/cancel the thread if it exists at the end of the method before it returns

Reply:
end of UCI::loop? I agree, yeah